internal/app/cache/cache.go

```diff
@@ -11,9 +17,128 @@ type Cache interface {
 	// Fetch returns the cached response if it exists.
 	Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error)
 
-	// SetResponse sets the cache response and returns the list of open watches.
+	// SetResponse sets the cache response and returns the list of requests.
 	SetResponse(key string, resp envoy_api_v2.DiscoveryResponse) ([]*envoy_api_v2.DiscoveryRequest, error)
 
-	// AddWatch adds the watch to the cache and returns whether a stream is open.
-	AddWatch(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
+	// AddRequest adds the request to the cache and returns whether a stream is open.
+	AddRequest(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
+}
+
+type cache struct {
+	cacheMu sync.RWMutex
+	cache   *ristretto.Cache
+	ttl     time.Duration
+}
+
+type resource struct {
+	resp       *envoy_api_v2.DiscoveryResponse
+	requests   []*envoy_api_v2.DiscoveryRequest
+	streamOpen bool
+}
+
+// Callback function for each eviction. Receives the key hash, conflict hash, cache value, and cost when called.
+type onEvictFunc func(key uint64, conflict uint64, value interface{}, cost int64)
+
+func NewCache(numCounters int64, cacheSizeBytes int64, ttl time.Duration, onEvict onEvictFunc) (Cache, error) {
+	// Config values are set as recommended in the ristretto documentation: https://github.com/dgraph-io/ristretto#Config.
+	config := ristretto.Config{
+		// NumCounters sets the number of counters/keys to keep for tracking access frequency.
+		// The value is recommended to be higher than the max cache capacity,
+		// and 10x the number of unique items in the cache when full (note that this is different from MaxCost).
+		NumCounters: numCounters,
+		// MaxCost denotes the max size in bytes and determines how eviction decisions are made.
+		// For consistency, the cost passed into SetWithTTL represents the size of the value in bytes.
+		MaxCost: cacheSizeBytes,
+		// BufferItems is the size of the Get buffers. The recommended value is 64.
+		BufferItems: 64,
+		// OnEvict is called for each eviction and closes the stream if a key is removed (e.g. expiry due to TTL).
+		OnEvict: onEvict,
+	}
+	newCache, err := ristretto.NewCache(&config)
+	if err != nil {
+		return nil, err
+	}
+	return &cache{
+		cache: newCache,
+		ttl:   ttl,
+	}, nil
+}
+
+func (c *cache) Exists(key string) bool {
+	_, found := c.cache.Get(key)
```

Comment:
We should probably check for whether the response exists for the key rather than if any value exists.

Originally I imagined that the orchestrator would call this prior to `Fetch`, but thinking more we could probably remove this func altogether in favor of just nil checking on `Fetch`.

Can you think of other use cases where we might want to call this `Exists`?

Reply:
cc @jyotimahapatra

Reply:
SGTM

Reply:
Do you mean the use case will use nil checking on https://github.com/envoyproxy/xds-relay/pull/25/files#diff-daf2807d54a5369eb2a29014b0d051cfR81 ?

Reply:
Agreed, you can just check if the return value from Fetch is nil if that was the intent.

Reply:
FWIW it was useful for test validation